### PR TITLE
[BEAM-115] Runner API representation of windowing strategies for Python

### PR DESCRIPTION
**Pull request by robertwb**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by robertwb** [comment]
R: @vikkyrk

**Comment by kennknowles** [comment]
Is the plan to check in the generated code?

File: sdks/python/apache_beam/pipeline.py [hunk 1]
```diff
@@ -440,3 +443,62 @@ def visit(self, visitor, pipeline, visited):
...
+    self._pipeline_context = context
+    self._obj_type = obj_type
+    self._obj_to_id = {}
+    self._id_to_obj = {}
+    self._id_to_proto = proto_map if proto_map else {}
+    self._counter = 0
+
+  def _unique_ref(self):
+    self._counter += 1
+    return "ref_%s_%s" % (obj_type.__name__, self._counter)
```
Comment (tibkiss) [inline sdks/python/apache_beam/pipeline.py]: obj_type is not available in this function, please use self._obj_type

File: sdks/python/apache_beam/pipeline.py [hunk 2]
```diff
@@ -440,3 +443,62 @@ def visit(self, visitor, pipeline, visited):
...
+  }
+
+  def __init__(self, context_proto=None):
+    for name, cls in self._COMPONENT_TYPES.items():
+      setattr(self, name,
+              PipelineContextMap(self, cls, getattr(context_proto, name, None)))
+
+  def to_runner_api(self):
+    context_proto = beam_runner_api_pb2.Components()
+    for name, cls in self.__COMPONENT_TYEPS:
```
Comment (tibkiss) [inline sdks/python/apache_beam/pipeline.py]: typo in self.__COMPONENT_TYEPS

File: sdks/python/apache_beam/pipeline.py [hunk 3]
```diff
@@ -440,3 +443,62 @@ def visit(self, visitor, pipeline, visited):
...
+      getattr(self, name).populate_map(getattr(context_proto, name))
+    return components
```
Comment (tibkiss) [inline sdks/python/apache_beam/pipeline.py]: Seems like components is undefined at this point

File: sdks/python/apache_beam/transforms/core.py [hunk 4]
```diff
@@ -1210,6 +1211,30 @@ def __repr__(self):
...
+            self.window_fn.get_window_coder()),
+        trigger=self.triggerfn.to_runner_api(context),
+        accumulation_mode=self.accumulation_mode,
+        output_time=self.output_time,
+        closing_behavior=EMIT_ALWAYS,
+        allowed_lateness=0)
+
+  @staticmethod
+  def from_runner_api(self, proto, context):
+    return Windowing(
```
Comment (tibkiss) [inline sdks/python/apache_beam/transforms/core.py]: Windowing class has different named parameters: windowfn and triggerfn is used there (without underscore).

File: sdks/python/apache_beam/transforms/core.py [hunk 5]
```diff
@@ -1210,6 +1211,30 @@ def __repr__(self):
...
+        window_fn=WindowFn.from_runner_api(proto.window_fn),
+        trigger_fn=TriggerFn.from_runner_api(proto.trigger),
```
Comment (tibkiss) [inline sdks/python/apache_beam/transforms/core.py]: TriggerFn was not imported in this file

File: sdks/python/apache_beam/transforms/core.py [hunk 6]
```diff
@@ -1210,6 +1211,30 @@ def __repr__(self):
...
+        window_coder_id=context.coders.get_id(
+            self.window_fn.get_window_coder()),
+        trigger=self.triggerfn.to_runner_api(context),
+        accumulation_mode=self.accumulation_mode,
+        output_time=self.output_time,
+        closing_behavior=EMIT_ALWAYS,
+        allowed_lateness=0)
+
+  @staticmethod
+  def from_runner_api(self, proto, context):
```
Comment (tibkiss) [inline sdks/python/apache_beam/transforms/core.py]: A staticmethod shall not need 'self' parameter

File: sdks/python/apache_beam/transforms/trigger.py [hunk 7]
```diff
@@ -474,12 +581,37 @@ def reset(self, window, context):
...
+  def from_runner_api(proto, context):
+    return AfterEach([
+      TriggerFn.from_runner_api(subtrigger, context)
+      for subtrigger in proto.after_each.subtriggers])
+
+  def to_runner_api(self, context):
+    return beam_runner_api_pb2.Trigger(
+        after_each=beam_runner_api_pb2.Trigger.AfterEach(
+            subtriggers=[
+                subtrigger.to_runner_api(context) for subtrigger in triggers]))
```
Comment (tibkiss) [inline sdks/python/apache_beam/transforms/trigger.py]: typo: please replace triggers by self.triggers

File: sdks/python/apache_beam/transforms/trigger.py [hunk 8]
```diff
@@ -474,12 +581,37 @@ def reset(self, window, context):
...
+    return OrFinally(
+        TriggerFn.from_runner_api(proto.or_finally.main, context),
+        TriggerFn.from_runner_api(getattr(proto.or_finally, 'finally'),
+                                  context))
+
+  def to_runner_api(self, context):
+    return beam_runner_api_pb2.Trigger(
+        or_finally=beam_runner_api_pb2.Trigger.OrFinally(
+            main=self.triggers[0].to_runner_api(context),
+            **{'finally': self.triggers[1].to_runner_api(context)}))
```
Comment (tibkiss) [inline sdks/python/apache_beam/transforms/trigger.py]: Maybe it's just me, but I needed some time to figure out what it does. 
Wouldn't be simpler just passing `finally=self.triggers[1].to_runner_api(context)`?

File: sdks/python/apache_beam/pipeline.py (hunk 1 above)
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: Done. I'm going to add tests for this.

File: sdks/python/apache_beam/pipeline.py (hunk 2 above)
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: Done.

File: sdks/python/apache_beam/transforms/core.py [hunk 9]
```diff
@@ -1210,6 +1211,30 @@ def __repr__(self):
   def is_default(self):
     return self._is_default
 
+  def to_runner_api(self, context):
+    raise NotImplementedError
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: Oh, yes.

File: sdks/python/apache_beam/pipeline.py (hunk 3 above)
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: Fixed.

File: sdks/python/apache_beam/transforms/core.py (hunk 4 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: Fixed.

File: sdks/python/apache_beam/transforms/core.py (hunk 5 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: Yes, there was a circular import. Fixed.

File: sdks/python/apache_beam/transforms/core.py (hunk 6 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: Done.

File: sdks/python/apache_beam/transforms/trigger.py (hunk 7 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/trigger.py]: Done.

File: sdks/python/apache_beam/transforms/trigger.py (hunk 8 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/trigger.py]: Yeah, if only finally wasn't a keyword in Python, so I can't put it literally.

**Comment by robertwb** [comment]
Thanks for taking a look, @tibkiss. I though I had added tests for Windowing, but clearly not.

File: sdks/python/apache_beam/pipeline.py (hunk 2 above)
Comment (tibkiss) [inline sdks/python/apache_beam/pipeline.py]: Thanks for addressing my comments!
There is still a typo here: *TYEPS vs. *TYPES

Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: Oh. Fixed. I'll add some tests to make sure this actually works an push again. (Oh, for a statically typechecked Python...)

**Comment by robertwb** [comment]
@kennknowles As for checking in generated code, I poked around a bit, and asked around, and that's what we're currently doing for the other protos and apitools generated code. It'd be nice to avoid a protoc/mvn dependency for contributors, but I figured this was better than generating the json manually.

**Comment by robertwb** [comment]
R: @kennknowles

**Comment by kennknowles** [comment]
@robertwb with regard to checking it in, I actually don't have much of an objection. It seems like a potentially annoying problem, and this unblocks other work. Perhaps we'll get to a point where it is autogenerated later, and it won't be hard to adjust other things.

I'll take a little bit to get to this review in more depth, though. I trust the LGTM of someone involved in the Python SDK more deeply.

File: sdks/python/apache_beam/transforms/core.py [hunk 10]
```diff
@@ -1207,9 +1209,45 @@ def __repr__(self):
...
+        # TODO(robertwb): Prohibit implicit multi-level merging.
+        merge_status=(beam_runner_api_pb2.NEEDS_MERGE
+                      if self.windowfn.is_merging()
+                      else beam_runner_api_pb2.NON_MERGING),
+        window_coder_id=context.coders.get_id(
+            self.windowfn.get_window_coder()),
+        trigger=self.triggerfn.to_runner_api(context),
+        accumulation_mode=self.accumulation_mode,
+        output_time=self.output_time_fn,
+        closing_behavior=beam_runner_api_pb2.EMIT_ALWAYS,
```
Comment (vikkyrk) [inline sdks/python/apache_beam/transforms/core.py]: Add a TODO to support `EMIT_IF_NONEMPTY`?

File: sdks/python/apache_beam/pipeline.py [hunk 11]
```diff
@@ -440,3 +443,62 @@ def visit(self, visitor, pipeline, visited):
         if v not in visited:
           visited.add(v)
           visitor.visit_value(v, self)
+
+
+class PipelineContextMap(object):
```
Comment (vikkyrk) [inline sdks/python/apache_beam/pipeline.py]: Tests for `PipelineContextMap` and `PipelineContext` would be great.

File: sdks/python/apache_beam/pipeline.py [hunk 12]
```diff
@@ -440,3 +443,62 @@ def visit(self, visitor, pipeline, visited):
...
+    return self._obj_to_id[obj]
+
+  def get_by_id(self, id):
+    if id not in self._id_to_obj:
+      self._id_to_obj[id] = self._obj_type.from_runner_api(
+        self._id_to_proto[id], self._pipeline_context)
+    return self._id_to_obj[id]
+
+
+class PipelineContext(object):
```
Comment (vikkyrk) [inline sdks/python/apache_beam/pipeline.py]: Is this a user facing class or should it it be in `runners` module.

File: sdks/python/apache_beam/utils/urns.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+# limitations under the License.
+#
+
+PICKLED_WINDOW_FN = "beam:window_fn:pickled_python:v0.1"
+GLOBAL_WINDOWS_FN = "beam:window_fn:global_windows:v0.1"
+FIXED_WINDOWS_FN = "beam:window_fn:fixed_windows:v0.1"
+SLIDING_WINDOWS_FN = "beam:window_fn:sliding_windows:v0.1"
+SESSION_WINDOWS_FN = "beam:window_fn:session_windows:v0.1"
+
+PICKLED_CODER = "dataflow:coder:pickled_python:v0.1"
```
Comment (robertwb) [inline sdks/python/apache_beam/utils/urns.py]: Done.

File: sdks/python/apache_beam/transforms/window.py [hunk 14]
```diff
@@ -49,23 +49,31 @@
...
 class OutputTimeFn(object):
   """Determines how output timestamps of grouping operations are assigned."""
 
-  OUTPUT_AT_EOW = 'OUTPUT_AT_EOW'
-  OUTPUT_AT_EARLIEST = 'OUTPUT_AT_EARLIEST'
-  OUTPUT_AT_LATEST = 'OUTPUT_AT_LATEST'
+  OUTPUT_AT_EOW = beam_runner_api_pb2.END_OF_WINDOW
+  OUTPUT_AT_EARLIEST = beam_runner_api_pb2.EARLIEST_IN_PANE
+  OUTPUT_AT_LATEST = beam_runner_api_pb2.LATEST_IN_PANE
   OUTPUT_AT_EARLIEST_TRANSFORMED = 'OUTPUT_AT_EARLIEST_TRANSFORMED'
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/window.py]: Done.

File: sdks/python/apache_beam/transforms/window.py [hunk 15]
```diff
@@ -132,6 +144,42 @@ def get_transformed_output_time(self, window, input_timestamp):  # pylint: disab
...
+        spec=beam_runner_api_pb2.UrnWithParameter(
+            urn=urn,
+            parameter=proto_utils.pack_Any(typed_param)))
+
+  @staticmethod
+  def from_runner_api_parameter(fn_parameter, unused_context):
+    return pickler.loads(fn_parameter.value)
+
+  def to_runner_api_parameter(self, context):
+    raise TypeError(self)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/window.py]: Done.

File: sdks/python/apache_beam/transforms/core.py (hunk 10 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: Done.

File: sdks/python/apache_beam/pipeline.py (hunk 11 above)
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: Done.

File: sdks/python/apache_beam/pipeline.py (hunk 12 above)
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: Done.

**Comment by vikkyrk** [comment]
LGTM.

**Pull request closed without merging**